notebooks/20_markov_decision_processes_part_2/index.md

```diff
@@ -0,0 +1,672 @@
+
+<br/>
+
+**function** Q-Value(MDP, $s, a, U$) **returns** a utility value <br/>
+&emsp; $\Sigma_{s'} {P(s^{\prime}|s,a)} \Big [R(s,a,s') + {\gamma}U^{\star}(s') \Big]$
+
+<br/>
+
+**function** Policy-Evaluation($\pi, U$, MDP) **returns** a utility function <br/>
+&emsp; **inputs**: MDP, an MDP with states $S$ , actions $A(s)$, transition model $P(s^{\prime}|s,a)$,     
+&emsp; &emsp; &emsp; &emsp; &emsp; rewards $R(s,a,s^{\prime})$, discount $\gamma$     
+&emsp; &emsp; &emsp; &emsp; &emsp; $U$, $U^{\prime}$, utility functions for states in S with policy $\pi$   
+&emsp; &emsp; &emsp; &emsp; &emsp; $\delta$, the maximum relative change in the utility of any state      
+&emsp; **repeat**       
+&emsp; &emsp; &emsp; $U \leftarrow U^{\prime}$;$\delta \leftarrow 0$       
+&emsp; &emsp; &emsp;**for each** state s **in** S **do**       
+&emsp; &emsp; &emsp; &emsp; $U^{\prime}(s) \leftarrow \sum_{s^\prime}^{} P(s^{\prime}|s,a)[R(s,a,s^{\prime})+\gamma U(s^{\prime})]$      
+&emsp; &emsp; &emsp; &emsp; **if** &nbsp; $\lvert U^{\prime}(s) - U(s)\lvert > \delta$ &nbsp; **then** &nbsp;  $\delta \leftarrow \lvert U^{\prime}(s) - U(s)\lvert$      
+&emsp; **until** $\delta \leq \frac{1 - \gamma}{\gamma} \epsilon$      
+&emsp; **return** $U$
+
+<br/>
+
+**function** Policy-Iteration(MDP) **returns** a utility function     
+&emsp; **inputs**: MDP, an MDP with states $S$ , actions $A(s)$, transition model $P(s^{\prime}|s,a)$,     
+&emsp; &emsp; &emsp; &emsp; $U$ , a vector of utilities for states in $S$, initially zero      
+&emsp; &emsp; &emsp;&emsp; &emsp; &emsp; &emsp; &emsp; $\pi$, a policy vector indexed by state, initially random     
+&emsp; **repeat**      
+&emsp; &emsp; $U \leftarrow $ Policy-Evaluation($\pi$, $U$, MDP)      
+&emsp; &emsp; $unchanged? \leftarrow$ true     
+&emsp; &emsp; **for each** state s **in** S **do** <br/>
+&emsp; &emsp; &emsp; $a^{\star} \leftarrow \underset{a \in A(s)}{\operatorname{argmax}}$ Q-Value(MDP, $s, a, U$)<br/>
+&emsp; &emsp; &emsp; **if** &nbsp; $\max\limits_{a \in A(s)}$ Q-Value(MDP, $s, a^{\star}, U$) $>$ Q-Value(MDP, $s, \pi[s], U$) &nbsp; **then** &nbsp;  **do**      
+&emsp; &emsp; &emsp; &emsp; $\pi[s] \leftarrow a^{\star}$     
+&emsp; &emsp; &emsp; &emsp; $unchanged? \leftarrow$ false     
+&emsp; **until** $unchanged?$     
+&emsp; **return** $\pi$
+
+
+**Optimality:**
+Policy iteration is optimal
+
+**Convergence:**
+Can converge mush faster under some condition
+
+**Efficiency:**
+$O(\lvert S\rvert^{2} \lvert A\rvert)=O(\lvert S\rvert^{2})+O(\lvert S\rvert^{2} \lvert A\rvert)$ which are orders of policy evaluation and policy improvement. 
+
+<br/>
+
+<div id="conclusion"><h1> Conclusion </h1></div>
+---
+<br/>
+
+These two methods are compared and summarized in this section. In general, both are kinds of dynamic programming algorithms and guarantee convergence. As you see, in both you must use Bellman equations.
```

Comment:
In a new section, you should mention their name instead of referring them by "These two"